crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+use std::iter::FusedIterator;
+
+use ruff_db::file_system::{FileSystemPath, FileSystemPathBuf};
+use ruff_db::vfs::{system_path_to_file, VfsPath};
+
+use crate::module_name::ModuleName;
+use crate::supported_py_version::get_target_py_version;
+use crate::typeshed::{parse_typeshed_versions, TypeshedVersionsQueryResult};
+use crate::{Db, TypeshedVersions};
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct ExtraPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct ExtraPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct FirstPartyPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct FirstPartyPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct StandardLibraryPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct StandardLibraryPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct SitePackagesPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct SitePackagesPathBuf(FileSystemPathBuf);
+
+/// Enumeration of the different kinds of search paths type checkers are expected to support.
+///
+/// N.B. Although we don't implement `Ord` for this enum, they are ordered in terms of the
+/// priority that we want to give these modules when resolving them,
+/// as per [the order given in the typing spec]
+///
+/// [the order given in the typing spec]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) enum ModuleResolutionPath {
+    Extra(ExtraPathBuf),
+    FirstParty(FirstPartyPathBuf),
+    StandardLibrary(StandardLibraryPathBuf),
+    SitePackages(SitePackagesPathBuf),
+}
+
+impl ModuleResolutionPath {
+    /// Push a new part to the path,
+    /// while maintaining the invariant that the path can only have `.py` or `.pyi` extensions.
+    /// For the stdlib variant specifically, it may only have a `.pyi` extension.
+    ///
+    /// ## Panics:
+    /// If a component with an invalid extension is passed
+    pub(crate) fn push(&mut self, component: &str) {
+        debug_assert!(matches!(component.matches('.').count(), 0 | 1));
+        if cfg!(debug) {
+            if let Some(extension) = std::path::Path::new(component).extension() {
```

Comment:
```suggestion
            if let Some(extension) = camino::Utf8Path::new(component).extension() {
```

since you already depend on `Utf8Path` anyway and it avoids having to call `to_str` to deal with non unicode paths